nannou_wgpu/src/texture/image.rs

```diff
@@ -564,14 +614,71 @@ where
+    let first_buffer = buffers.next()?;
+
+    // Build the texture ready to receive the data.
+    let (width, height) = first_buffer.dimensions();
+    let extent = wgpu::Extent3d {
+        width,
+        height,
+        depth_or_array_layers: array_layers,
     };
-    let mut encoder = device.create_command_encoder(&cmd_encoder_desc);
-    let texture =
-        encode_load_texture_array_from_image_buffers(device, &mut encoder, usage, buffers);
-    queue.submit(std::iter::once(encoder.finish()));
-    texture
+    let texture = wgpu::TextureBuilder::from_image_view(first_buffer)
+        .extent(extent)
+        .dimension(wgpu::TextureDimension::D2) // force an array
+        .usage(wgpu::TextureBuilder::REQUIRED_IMAGE_TEXTURE_USAGE | usage)
+        .build(device);
+
+    // Describe the layout of the data.
+    let format = texture.format();
+    let block_size = format.describe().block_size;
+    let bytes_per_row = extent.width * block_size as u32;
+    let padding = wgpu::compute_row_padding(bytes_per_row);
+    let padded_bytes_per_row = bytes_per_row + padding;
+    let image_data_layout = wgpu::ImageDataLayout {
+        offset: 0,
+        bytes_per_row: std::num::NonZeroU32::new(padded_bytes_per_row),
+        rows_per_image: std::num::NonZeroU32::new(height),
+    };
+
+    // Collect the data into a single slice, padding each row as necessary.
+    //
+    // NOTE: Previously we used `encode_load_texture_array_from_image_buffers` which avoids
+    // collecting the image data into a single slice. However, the `wgpu::Texture::from_*`
+    // constructors have been changed to avoid submitting an extra command buffer in favour
+    // of using `Queue::write_texture` which schedules the write for the next call to
+    // `Queue::submit`. This is to avoid an Intel driver bug where submitting more than one command
+    // buffer per frame appears to be causing issues:
+    // https://github.com/gfx-rs/wgpu/issues/1672#issuecomment-917510810
+    //
+    // While this likely means consuming more RAM, it also likely results in slightly better
+    // performance due to reducing the number of command buffers submitted.
+    //
+    // Users can still use `encode_load_texture_array_from_image_buffers` directly if they wish.
+    let capacity = bytes_per_row as usize * height as usize * array_layers as usize;
+    let mut data: Vec<u8> = Vec::with_capacity(capacity);
+    for buffer in Some(first_buffer).into_iter().chain(buffers) {
+        let layer_data = unsafe { wgpu::bytes::from_slice(&*buffer) };
+        for row_data in layer_data.chunks(bytes_per_row as usize) {
+            data.extend_from_slice(row_data);
+            data.extend((0..padding).map(|_| 0u8));
+        }
+    }
+
+    // Copy into the entire texture.
+    let image_copy_texture = wgpu::ImageCopyTexture {
+        texture: texture.inner(),
+        mip_level: 0,
+        origin: wgpu::Origin3d::ZERO,
+        // TODO: Maybe we shouldn't assume this?
```

Comment:
A quick way not to assume it might be `let image_copy_texture = texture.as_image_copy();`

Reply:
Agh of course, I forgot about that method, cheers!